app/services/earnings_csv_parser.rb

```diff
@@ -0,0 +1,83 @@
+require 'csv'
+require_relative '../models/earning'  # Add this line if Earning model is not being loaded
+
+class EarningsCsvParser
+  STRING = 'STRING'.freeze
+  DECIMAL = 'DECIMAL'.freeze
+  # file = CSV string
+  def initialize(file, employer_id)
+      @file = file
+      @employer = get_employer(employer_id)
+      @employer_csv_format = get_employer_csv_format
+  end
+
+  def create_earnings
+      csv = CSV.parse(@file, headers: true)
+      errors = []
+      csv.each do |row|
+        e = create_earning(row)
+        if e.valid?
+          e.save
+        else
+          errors << "Error with row #{row}: #{e.errors.full_messages}"
+        end
+      end
+      display_errors errors
+  end
+
+  private
+
+  def get_employer_csv_format
+    EmployerCsvFormat.find_by!(employer_id: @employer.id)
```

Comment:
In general, it's best to access associations through the auto-generate methods when you define the association.

Why? By using the methods you're encapsulating logic into the model. Right now this `EarningsCsvParser` needs to know both that an employer has_one format AND that it's set up as a separate table with a foreign key. If you use the ActiveRecord association method it gives you flexibility to change the underlying implementation without needing to update it here.